File: tools/music_audio/openai_audio/src/nodejs/openai_audio-tool.ts
import fs from 'node:fs'
import path from 'node:path'

import type { TranscriptionOutput } from '@tools/music_audio/transcription-schema'
import { Tool } from '@sdk/base-tool'
import { ToolkitConfig } from '@sdk/toolkit-config'
import { Network } from '@sdk/network'

// Hardcoded default settings for OpenAI audio tool
const OPENAI_AUDIO_API_KEY: string | null = null
const OPENAI_AUDIO_MODEL = 'whisper-1'
const DEFAULT_SETTINGS: Record<string, unknown> = {
  OPENAI_AUDIO_API_KEY,
  OPENAI_AUDIO_MODEL
}
const REQUIRED_SETTINGS = ['OPENAI_AUDIO_API_KEY']

interface OpenAITranscriptionOutput {
  task: string
  duration: number
  text: string
  segments: {
    type: string
    id: string
    start: number
    end: number
    text: string
    speaker: string
  }[]
  usage: {
    type: string
    seconds: number
  }
}

export default class OpenAIAudioTool extends Tool {
  private static readonly TOOLKIT = 'music_audio'
  private readonly config: ReturnType<typeof ToolkitConfig.load>
  readonly apiKey: string | null
  readonly model: string

  constructor() {
    super()
    this.config = ToolkitConfig.load(OpenAIAudioTool.TOOLKIT, this.toolName)

    const toolSettings = ToolkitConfig.loadToolSettings(
      OpenAIAudioTool.TOOLKIT,
      this.toolName,
      DEFAULT_SETTINGS
    )
    this.settings = toolSettings
    this.requiredSettings = REQUIRED_SETTINGS
    this.checkRequiredSettings(this.toolName)

    // Priority: toolkit settings > hardcoded default
    this.apiKey =
      (this.settings['OPENAI_AUDIO_API_KEY'] as string) || OPENAI_AUDIO_API_KEY
    this.model =
      (this.settings['OPENAI_AUDIO_MODEL'] as string) || OPENAI_AUDIO_MODEL
  }

  get toolName(): string {
    // Use the actual config name for toolkit lookup
    return 'openai_audio'
  }

  get toolkit(): string {
    return OpenAIAudioTool.TOOLKIT
  }

  get description(): string {
    return this.config['description']
  }

  /**
   * Transcribe audio to a file using OpenAI's audio transcription API via SDK Network
   * @param inputPath Path to the audio file to transcribe
   * @param outputPath Path to save the plain text transcription
   * @param apiKey OpenAI API key (uses env/hardcoded default if not provided)
   * @param model Transcription model (defaults to tool default)
   */
  async transcribeToFile(
    inputPath: string,
    outputPath: string,
    apiKey?: string,
    model?: string
  ): Promise<string> {
    // Use provided values, instance values, or error
    const finalApiKey = apiKey || this.apiKey
    const finalModel = model || this.model
    if (!finalApiKey) {
      throw new Error('OpenAI API key is missing')
    }

    const form = new FormData()
    const audioFile = await fs.openAsBlob(inputPath)
    form.append('file', audioFile, path.basename(inputPath))
    form.append('model', finalModel)
    form.append('chunking_strategy', 'auto')
    form.append('response_format', 'diarized_json')

    const network = new Network({ baseURL: 'https://api.openai.com' })
    const response = await network.request({
      url: '/v1/audio/transcriptions',
      method: 'POST',
      data: form,
      headers: {
        Authorization: `Bearer ${finalApiKey}`
      }
    })

    const parsedOutput = this.parseTranscription(
      response.data as OpenAITranscriptionOutput
    )

    await fs.promises.writeFile(
      outputPath,
      JSON.stringify(parsedOutput, null, 2),
      'utf8'
    )

    return outputPath
  }

  private parseTranscription(
    rawOutput: OpenAITranscriptionOutput
  ): TranscriptionOutput {
    const speakers = Array.from(
      new Set(rawOutput.segments.map((segment) => segment.speaker))
    )

    const segments = rawOutput.segments.map((segment) => {
      return {
        from: segment.start,
        to: segment.end,
        text: segment.text,
        speaker: segment.speaker || null
      }
    })

    // If duration is not found, use the "to" property from the last segment
    let duration = rawOutput.duration
    if (!duration && segments.length > 0) {
      duration = segments[segments.length - 1]?.to || 0
    }

    return {
      duration: duration || 0,
      speakers: speakers,
      speaker_count: speakers.length,
      segments,
      metadata: {
        tool: this.toolName
      }
    }
  }
}

File: tools/music_audio/openai_audio/src/python/openai_audio_tool.py
import json
from typing import List, Dict, Any, Optional

from bridges.python.src.sdk.base_tool import BaseTool
from bridges.python.src.sdk.toolkit_config import ToolkitConfig
from bridges.python.src.sdk.network import Network
from tools.music_audio.transcription_schema import TranscriptionOutput, TranscriptionSegment

# Hardcoded default settings for OpenAI audio tool
OPENAI_AUDIO_API_KEY = None
OPENAI_AUDIO_MODEL = "whisper-1"
DEFAULT_SETTINGS = {
    "OPENAI_AUDIO_API_KEY": OPENAI_AUDIO_API_KEY,
    "OPENAI_AUDIO_MODEL": OPENAI_AUDIO_MODEL,
}
REQUIRED_SETTINGS = ["OPENAI_AUDIO_API_KEY"]


class OpenAIAudioTool(BaseTool):
    TOOLKIT = "music_audio"

    def __init__(self):
        super().__init__()
        self.config = ToolkitConfig.load(self.TOOLKIT, self.tool_name)

        tool_settings = ToolkitConfig.load_tool_settings(
            self.TOOLKIT, self.tool_name, DEFAULT_SETTINGS
        )
        self.settings = tool_settings
        self.required_settings = REQUIRED_SETTINGS
        self._check_required_settings(self.tool_name)

        # Priority: toolkit settings > hardcoded default
        self.api_key = self.settings.get("OPENAI_AUDIO_API_KEY", OPENAI_AUDIO_API_KEY)
        self.model = self.settings.get("OPENAI_AUDIO_MODEL", OPENAI_AUDIO_MODEL)

        self.network = Network({"base_url": "https://api.openai.com"})

    @property
    def tool_name(self) -> str:
        # Use the actual config name for toolkit lookup
        return "openai_audio"

    @property
    def toolkit(self) -> str:
        return self.TOOLKIT

    @property
    def description(self) -> str:
        return self.config["description"]

    def transcribe_to_file(
        self,
        input_path: str,
        output_path: str,
        api_key: Optional[str] = None,
        model: Optional[str] = None,
    ) -> str:
        """
        Transcribe audio to a file using OpenAI's audio transcription API via SDK Network

        Args:
            input_path: Path to the audio file to transcribe
            output_path: Path to save the JSON transcription (unified format)
            api_key: OpenAI API key (uses env/hardcoded default if not provided)
            model: Transcription model (defaults to tool default)

        Returns:
            The path to the transcription file
        """
        # Use provided values, instance values, or error
        api_key = api_key or self.api_key
        model = model or self.model
        if not api_key:
            raise Exception("OpenAI API key is missing")

        try:
            files: dict = {"file": open(input_path, "rb")}
            data: dict = {
                "model": model,
                "chunking_strategy": "auto",
                "response_format": "diarized_json",
            }

            response = self.network.request(
                {
                    "url": "/v1/audio/transcriptions",
                    "method": "POST",
                    "headers": {"Authorization": f"Bearer {api_key}"},
                    "data": data,
                    "files": files,
                    "use_json": True,
                }
            )

            parsed_output = self._parse_transcription(response["data"])

            with open(output_path, "w", encoding="utf-8") as f:
                json.dump(parsed_output, f, indent=2, ensure_ascii=False)

            return output_path
        except Exception as e:
            raise Exception(f"OpenAI transcription failed: {str(e)}")

    def _parse_transcription(self, raw_output: Dict[str, Any]) -> TranscriptionOutput:
        segments_data = raw_output.get("segments", [])
        unique_speakers = list(
            set(seg.get("speaker") for seg in segments_data if seg.get("speaker"))
        )

        segments: List[TranscriptionSegment] = []
        for segment in segments_data:
            segments.append(
                {
                    "from": float(segment.get("start", 0)),
                    "to": float(segment.get("end", 0)),
                    "text": segment.get("text", ""),
                    "speaker": segment.get("speaker") or None,
                }
            )

        # If duration is not found, use the "to" property from the last segment
        duration = raw_output.get("duration")
        if not duration and len(segments) > 0:
            duration = segments[-1]["to"] or 0.0

        return {
            "duration": float(duration) if duration else 0.0,
            "speakers": unique_speakers,
            "speaker_count": len(unique_speakers),
            "segments": segments,
            "metadata": {"tool": self.tool_name},
        }
